src/Microsoft.AspNet.Razor/TagHelpers/TagHelperDescriptor.cs

```diff
@@ -154,13 +58,13 @@ public class TagHelperDescriptor
         /// <remarks>
         /// <c>*</c> at the end of an attribute name acts as a prefix match.
         /// </remarks>
-        public IReadOnlyList<string> RequiredAttributes { get; }
+        public IEnumerable<string> RequiredAttributes { get; [param: NotNull] set; } = Enumerable.Empty<string>();
 
         /// <summary>
         /// HTML element tag names targeted by tag helpers allowed as children.
         /// </summary>
         /// <remarks><c>null</c> indicates all tag helper children are allowed.</remarks>
-        public IReadOnlyList<string> AllowedChildren { get; }
+        public IEnumerable<string> AllowedChildren { get; set; }
```

Comment:
@ajaybhargavb ignore the specific collection types used and focus on the separation between these descriptors and the calling code. Each `TagHelperDescriptor` had unique instances of all exposed collections. We had this isolation in place because it's the normal guarantee for classes exposing collections i.e. they "own" the collection instances.

Now for example each instance of this class that `TagHelperDescriptorResolver` creates from its cache shares the collections with the original. We might get away with it if neither `TagHelperDescriptorResolver` nor `TagHelperDescriptorFactory` reuse the collection instances they're setting these properties to but it's not a great policy.

Double-check the code that creates `TagHelperDescriptor` instances to confirm we're :ok:.